File: my-app/backend/routes/messages.js
const express = require('express');
const router = express.Router();
const mongoose = require('mongoose');

const { body, validationResult } = require('express-validator');

const Message = require('../models/Message');
const Admin = require('../models/Admin');

/**
 * Get staffId linked to admin
 */
async function getExcludedStaffId(req) {
  try {
    const adminId = req.query.adminId;

    if (!adminId) return null;

    const admin = await Admin.findById(adminId).select('staffId');

    if (admin && admin.staffId) {
      return admin.staffId.toString();
    }

    return null;
  } catch (err) {
    console.error('Error resolving staffId:', err);
    return null;
  }
}

/**
 * Populate helper
 */
function withReportedStaffPopulated(query) {
  return query.populate('reportedStaff', 'name role');
}

/**
 * HARD FILTER (final protection)
 */
function filterOutOwnReports(messages, excludedStaffId) {
  if (!excludedStaffId) return messages;

  return messages.filter(msg => {
    if (!msg.reportedStaff) return true;

    return msg.reportedStaff._id.toString() !== excludedStaffId;
  });
}

/**
 * BUILD SAFE QUERY
 */
function buildSafeQuery(baseQuery, excludedStaffId) {
  if (!excludedStaffId) return baseQuery;

  return {
    ...baseQuery,
    $or: [
      { reportedStaff: null }, // ✅ include null
      {
        reportedStaff: {
          $ne: new mongoose.Types.ObjectId(excludedStaffId)
        }
      }
    ]
  };
}

/**
 * GET ALL MESSAGES
 */
router.get('/', async (req, res) => {
  try {
    const { threadToken, filter, date } = req.query;

    let baseQuery = { isDeleted: false };

    const excludedStaffId = await getExcludedStaffId(req);

    if (threadToken) {
      baseQuery.threadToken = threadToken;
    }

    if (date) {
      const start = new Date(date);
      const end = new Date(start);
      end.setDate(end.getDate() + 1);
      baseQuery.createdAt = { $gte: start, $lt: end };
    }

    if (filter) {
      const now = new Date();
      let startDate;

      if (filter === 'today') {
        startDate = new Date(now.getFullYear(), now.getMonth(), now.getDate());
      } else if (filter === 'last7') {
        startDate = new Date();
        startDate.setDate(startDate.getDate() - 7);
      } else if (filter === 'last30') {
        startDate = new Date();
        startDate.setDate(startDate.getDate() - 30);
      }

      if (startDate) {
        baseQuery.createdAt = { $gte: startDate };
      }
    }

    const query = buildSafeQuery(baseQuery, excludedStaffId);

    let messages = await withReportedStaffPopulated(
      Message.find(query).sort({ createdAt: -1 })
    );

    messages = filterOutOwnReports(messages, excludedStaffId);

    return res.json({
      success: true,
      messages: messages.map(m => ({
        ...m.toJSON(),
        id: m._id.toString()
      }))
    });

  } catch (err) {
    console.error('GET messages error:', err);
    res.status(500).json({ error: 'Failed to load messages.' });
  }
});

/**
 * GET UNREAD
 */
router.get('/unread', async (req, res) => {
  try {
    let baseQuery = { isRead: false, isDeleted: false };

    const excludedStaffId = await getExcludedStaffId(req);

    const query = buildSafeQuery(baseQuery, excludedStaffId);

    let messages = await withReportedStaffPopulated(
      Message.find(query).sort({ createdAt: -1 })
    );

    messages = filterOutOwnReports(messages, excludedStaffId);

    res.json({
      success: true,
      messages: messages.map(m => ({
        ...m.toJSON(),
        id: m._id.toString()
      }))
    });

  } catch (err) {
    console.error('Unread error:', err);
    res.status(500).json({ error: 'Failed to load unread messages.' });
  }
});

/**
 * GET ONE MESSAGE
 */
router.get('/:id', async (req, res) => {
  try {
    const excludedStaffId = await getExcludedStaffId(req);

    let message = await withReportedStaffPopulated(
      Message.findById(req.params.id)
    );

    if (!message || message.isDeleted) {
      return res.status(404).json({ error: 'Message not found.' });
    }

    if (
      excludedStaffId &&
      message.reportedStaff &&
      message.reportedStaff._id.toString() === excludedStaffId
    ) {
      return res.status(404).json({ error: 'Message not found.' });
    }

    res.json({
      success: true,
      message: {
        ...message.toJSON(),
        id: message._id.toString()
      }
    });

  } catch (err) {
    console.error('Get one error:', err);
    res.status(500).json({ error: 'Failed to load message.' });
  }
});

/**
 * MARK ALL UNREAD MESSAGES AS READ
 */
router.patch('/mark-all-read', async (req, res) => {
  try {
    let baseQuery = { isRead: false, isDeleted: false };
    const excludedStaffId = await getExcludedStaffId(req);
    const query = buildSafeQuery(baseQuery, excludedStaffId);

    const result = await Message.updateMany(query, { isRead: true, readAt: new Date() });
    const modified = result.modifiedCount ?? result.nModified ?? 0;

    return res.json({
      success: true,
      modifiedCount: modified
    });
  } catch (err) {
    console.error('Mark all as read error:', err);
    res.status(500).json({ error: 'Failed to mark messages as read.' });
  }
});

/**
 * MARK ONE MESSAGE AS READ
 */
router.patch('/:id/read', async (req, res) => {
  try {
    const excludedStaffId = await getExcludedStaffId(req);
    if (excludedStaffId) {
      const existing = await Message.findById(req.params.id).select('reportedStaff');
      if (existing && existing.reportedStaff && existing.reportedStaff.toString() === excludedStaffId) {
        return res.status(404).json({ error: 'Message not found.' });
      }
    }

    const updatedMessage = await withReportedStaffPopulated(
      Message.findByIdAndUpdate(
        req.params.id,
        { isRead: true, readAt: new Date() },
        { new: true }
      )
    );

    if (!updatedMessage) {
      return res.status(404).json({ error: 'Message not found.' });
    }

    res.json({
      success: true,
      message: {
        ...updatedMessage.toJSON(),
        id: updatedMessage._id.toString()
      }
    });

  } catch (err) {
    console.error('Mark read error:', err);
    res.status(500).json({ error: 'Failed to mark message as read.' });
  }
});

/**
 * CREATE MESSAGE
 */
router.post(
  '/',
  [
    body('topic').notEmpty(),
    body('message').notEmpty(),
    body('reportedStaff').optional().isMongoId(),
  ],
  async (req, res) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(422).json({ errors: errors.array() });
      }

      const { topic, message, reportedStaff } = req.body;

      const newMessage = new Message({
        topic,
        message,
        reportedStaff: reportedStaff || null,
      });

      await newMessage.save();

      res.status(201).json({
        success: true,
        threadToken: newMessage.threadToken
      });

    } catch (err) {
      console.error('Create error:', err);
      res.status(500).json({ error: 'Failed to send message.' });
    }
  }
);

module.exports = router;

File: my-app/backend-go/internal/app/messages.go
package app

import (
	"database/sql"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"
)

func messagesRoutes(store *Store) http.Handler {
	db := store.DB()
	r := chi.NewRouter()

	r.Get("/", func(w http.ResponseWriter, r *http.Request) {
		filter := r.URL.Query().Get("filter")
		date := r.URL.Query().Get("date")
		threadToken := r.URL.Query().Get("threadToken")
		excludedStaffID := r.URL.Query().Get("staffId")

		query := "SELECT id, topic, message, reportedStaff, threadToken, isRead, isDeleted, readAt, createdAt FROM messages WHERE isDeleted = 0"
		args := []any{}
		if threadToken != "" {
			query += " AND threadToken = ?"
			args = append(args, threadToken)
		}
		if date != "" {
			start, end, err := parseTimeRange(date)
			if err == nil {
				query += " AND createdAt >= ? AND createdAt < ?"
				args = append(args, start, end)
			}
		}
		if filter != "" {
			now := time.Now().UTC()
			var start time.Time
			switch filter {
			case "today":
				start = time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, time.UTC)
			case "last7":
				start = now.AddDate(0, 0, -7)
			case "last30":
				start = now.AddDate(0, 0, -30)
			}
			if !start.IsZero() {
				query += " AND createdAt >= ?"
				args = append(args, start)
			}
		}
		if excludedStaffID != "" {
			query += " AND (reportedStaff IS NULL OR reportedStaff <> ?)"
			args = append(args, excludedStaffID)
		}

		rows, err := db.QueryContext(r.Context(), query, args...)
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]any{"error": "Failed to load messages."})
			return
		}
		defer rows.Close()
		var messages []Message
		for rows.Next() {
			m, ok := scanMessage(rows)
			if !ok {
				continue
			}
			messages = append(messages, m)
		}
		if err := rows.Err(); err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]any{"error": "Failed to load messages."})
			return
		}
		publicMessages := make([]map[string]any, 0, len(messages))
		for _, msg := range messages {
			publicMessages = append(publicMessages, msg.public())
		}
		writeJSON(w, http.StatusOK, map[string]any{"success": true, "messages": publicMessages})
	})

	r.Get("/unread", func(w http.ResponseWriter, r *http.Request) {
		query := "SELECT id, topic, message, reportedStaff, threadToken, isRead, isDeleted, readAt, createdAt FROM messages WHERE isRead = 0 AND isDeleted = 0"
		args := []any{}
		if excludedStaffID := r.URL.Query().Get("staffId"); excludedStaffID != "" {
			query += " AND (reportedStaff IS NULL OR reportedStaff <> ?)"
			args = append(args, excludedStaffID)
		}
		rows, err := db.QueryContext(r.Context(), query, args...)
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]any{"error": "Failed to load unread messages."})
			return
		}
		defer rows.Close()
		var messages []Message
		for rows.Next() {
			m, ok := scanMessage(rows)
			if !ok {
				continue
			}
			messages = append(messages, m)
		}
		if err := rows.Err(); err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]any{"error": "Failed to load unread messages."})
			return
		}
		publicMessages := make([]map[string]any, 0, len(messages))
		for _, msg := range messages {
			publicMessages = append(publicMessages, msg.public())
		}
		writeJSON(w, http.StatusOK, map[string]any{"success": true, "messages": publicMessages})
	})

	r.Get("/{id}", func(w http.ResponseWriter, r *http.Request) {
		id := chi.URLParam(r, "id")
		num, err := strconv.ParseInt(id, 10, 64)
		if err != nil || num == 0 {
			writeJSON(w, http.StatusNotFound, map[string]any{"error": "Message not found."})
			return
		}
		row := db.QueryRowContext(r.Context(), "SELECT id, topic, message, reportedStaff, threadToken, isRead, isDeleted, readAt, createdAt FROM messages WHERE id = ? AND isDeleted = 0", num)
		msg, ok := scanMessage(row)
		if !ok {
			writeJSON(w, http.StatusNotFound, map[string]any{"error": "Message not found."})
			return
		}
		if excludedStaffID := r.URL.Query().Get("staffId"); excludedStaffID != "" && msg.ReportedStaff != nil && *msg.ReportedStaff == excludedStaffID {
			writeJSON(w, http.StatusNotFound, map[string]any{"error": "Message not found."})
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"success": true, "message": msg.public()})
	})

	r.Patch("/{id}/read", func(w http.ResponseWriter, r *http.Request) {
		id := chi.URLParam(r, "id")
		num, err := strconv.ParseInt(id, 10, 64)
		if err != nil || num == 0 {
			writeJSON(w, http.StatusNotFound, map[string]any{"error": "Message not found."})
			return
		}
		var current Message
		var reportedStaff sql.NullString
		var readAt sql.NullTime
		var isDeleted, isRead int
		err = db.QueryRowContext(r.Context(), "SELECT id, reportedStaff, isRead, isDeleted, readAt FROM messages WHERE id = ?", num).Scan(&current.ID, &reportedStaff, &isRead, &isDeleted, &readAt)
		if err != nil {
			writeJSON(w, http.StatusNotFound, map[string]any{"error": "Message not found."})
			return
		}
		if reportedStaff.Valid {
			current.ReportedStaff = &reportedStaff.String
		}
		if excludedStaffID := r.URL.Query().Get("staffId"); excludedStaffID != "" && current.ReportedStaff != nil && *current.ReportedStaff == excludedStaffID {
			writeJSON(w, http.StatusNotFound, map[string]any{"error": "Message not found."})
			return
		}
		if isDeleted != 0 {
			writeJSON(w, http.StatusNotFound, map[string]any{"error": "Message not found."})
			return
		}
		now := time.Now().UTC()
		if _, err := db.ExecContext(r.Context(), "UPDATE messages SET isRead = 1, readAt = ? WHERE id = ?", now, num); err != nil {
			writeJSON(w, http.StatusNotFound, map[string]any{"error": "Message not found."})
			return
		}
		current.IsRead = true
		current.IsDeleted = isDeleted != 0
		current.ReadAt = &now
		writeJSON(w, http.StatusOK, map[string]any{"success": true, "message": current.public()})
	})

	r.Patch("/mark-all-read", func(w http.ResponseWriter, r *http.Request) {
		query := "UPDATE messages SET isRead = 1, readAt = ? WHERE isRead = 0 AND isDeleted = 0"
		args := []any{time.Now().UTC()}
		if excludedStaffID := r.URL.Query().Get("staffId"); excludedStaffID != "" {
			query += " AND (reportedStaff IS NULL OR reportedStaff <> ?)"
			args = append(args, excludedStaffID)
		}
		result, err := db.ExecContext(r.Context(), query, args...)
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]any{"error": "Failed to mark messages as read."})
			return
		}
		modified, _ := result.RowsAffected()
		writeJSON(w, http.StatusOK, map[string]any{"success": true, "modifiedCount": modified})
	})

	r.Post("/", func(w http.ResponseWriter, r *http.Request) {
		var payload struct {
			Topic         string `json:"topic"`
			Message       string `json:"message"`
			ReportedStaff string `json:"reportedStaff"`
		}
		if err := decodePayload(r, &payload); err != nil {
			writeJSON(w, http.StatusBadRequest, map[string]any{"error": "Invalid JSON payload"})
			return
		}
		topic := strings.TrimSpace(payload.Topic)
		message := strings.TrimSpace(payload.Message)
		if topic == "" || message == "" {
			writeJSON(w, http.StatusBadRequest, map[string]any{"error": "Topic and message are required"})
			return
		}
		threadToken := randomHex(12)
		now := time.Now().UTC()
		var reportedStaff any
		if payload.ReportedStaff != "" {
			reportedStaff = payload.ReportedStaff
		}
		result, err := db.ExecContext(r.Context(),
			"INSERT INTO messages (topic, message, reportedStaff, threadToken, isRead, isDeleted, createdAt) VALUES (?, ?, ?, ?, 0, 0, ?)",
			topic, message, reportedStaff, threadToken, now)
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]any{"error": "Failed to send message."})
			return
		}
		id, _ := result.LastInsertId()
		msg := Message{ID: id, Topic: topic, Message: message, ThreadToken: threadToken, IsRead: false, IsDeleted: false, CreatedAt: now}
		if payload.ReportedStaff != "" {
			msg.ReportedStaff = &payload.ReportedStaff
		}
		writeJSON(w, http.StatusCreated, map[string]any{"success": true, "threadToken": msg.ThreadToken})
	})

	return r
}

// scanner abstracts *sql.Row and *sql.Rows for scanning a message row.
type scanner interface {
	Scan(dest ...any) error
}

func scanMessage(s scanner) (Message, bool) {
	var m Message
	var reportedStaff sql.NullString
	var readAt sql.NullTime
	var isRead, isDeleted int
	if err := s.Scan(&m.ID, &m.Topic, &m.Message, &reportedStaff, &m.ThreadToken, &isRead, &isDeleted, &readAt, &m.CreatedAt); err != nil {
		if err == sql.ErrNoRows {
			return m, false
		}
		return m, false
	}
	if reportedStaff.Valid {
		m.ReportedStaff = &reportedStaff.String
	}
	m.IsRead = isRead != 0
	m.IsDeleted = isDeleted != 0
	if readAt.Valid {
		t := readAt.Time.UTC()
		m.ReadAt = &t
	}
	m.CreatedAt = m.CreatedAt.UTC()
	return m, true
}
